Bottom: 14ebf5701bba2dc96fb1f29b02dee56748353c6f
Top:    6a3cc6d8f9f3245ba495aa10a17dc125519948f1
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-05-26 12:01:35 +0800

vhost_net: conditionally start tx polling

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 9b51989..686bc07 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -354,6 +354,7 @@ static int vhost_net_tx_get_vq_desc(struct vhost_net *net,
 	return r;
 }
 
+
 static bool vhost_exceeds_maxpend(struct vhost_net *net)
 {
 	struct vhost_net_virtqueue *nvq = &net->vqs[VHOST_NET_VQ_TX];
@@ -394,6 +395,7 @@ static void handle_tx(struct vhost_net *net)
 		goto out;
 
 	vhost_disable_notify(&net->dev, vq);
+	vhost_net_disable_vq(net, vq);
 
 	hdr_size = nvq->vhost_hlen;
 	zcopy = nvq->ubufs;
@@ -484,6 +486,8 @@ static void handle_tx(struct vhost_net *net)
 					% UIO_MAXIOV;
 			}
 			vhost_discard_vq_desc(vq, 1);
+			if (err == -EAGAIN)
+				vhost_net_enable_vq(net, vq);
 			break;
 		}
 		if (err != len)
